mp/src/game/client/momentum/c_mom_triggers.cpp

```diff
@@ -80,64 +109,167 @@ C_BaseMomZoneTrigger::C_BaseMomZoneTrigger()
+
+    CMatRenderContextPtr pRenderContext(materials);
+    IMesh *pMesh = pRenderContext->GetDynamicMesh(true, nullptr, nullptr, materials->FindMaterial("momentum/zone_outline", TEXTURE_GROUP_OTHER));
+    CMeshBuilder builder;
+
+    for (int i = 0; i < iNum; i++)
+    {
+        const auto& vecCurr = m_vecZonePoints[i];
+        const auto& vecNext = m_vecZonePoints[(i + 1) % iNum];
+
+        builder.Begin(pMesh, MATERIAL_QUADS, 4);
+
+        builder.Position3fv(vecCurr.Base());
+        builder.Color4ub(faceColor.r(), faceColor.g(), faceColor.b(), faceAlpha);
+        builder.AdvanceVertex();
+
+        builder.Position3f(vecCurr.x, vecCurr.y, vecCurr.z + m_flZoneHeight);
+        builder.Color4ub(faceColor.r(), faceColor.g(), faceColor.b(), faceAlpha);
+        builder.AdvanceVertex();
+
+        builder.Position3f(vecNext.x, vecNext.y, vecNext.z + m_flZoneHeight);
+        builder.Color4ub(faceColor.r(), faceColor.g(), faceColor.b(), faceAlpha);
+        builder.AdvanceVertex();
+
+        builder.Position3fv(vecNext.Base());
+        builder.Color4ub(faceColor.r(), faceColor.g(), faceColor.b(), faceAlpha);
+        builder.AdvanceVertex();
+        
+        builder.End(false, true);
+    }
+}
+
+void C_BaseMomZoneTrigger::DrawSideFacesModelAsOverlay()
+{
+    const int iNum = m_vecZonePoints.Count();
+    if (iNum <= 2)
+        return;
+
+    Color faceColor = m_ZoneModelRenderer.m_Color;
+
+    int faceAlpha = mom_zone_draw_alpha_override_toggle.GetBool() ? mom_zone_draw_faces_alpha_override.GetInt() : faceColor.a();
+
+    Vector min = Vector(0, 0, 0);
+    for (int i = 0; i < iNum; i++)
+    {
+        const auto& vecCurr = m_vecZonePoints[i];
+        const auto& vecNext = m_vecZonePoints[(i + 1) % iNum];
+
+        QAngle angle(0.0f, 0.0f, 0.0f);
+        float iWidth = sqrtf(Sqr(vecNext.x - vecCurr.x) + Sqr(vecNext.y - vecCurr.y));
+
+        VectorAngles(vecNext - vecCurr, angle);
+
+        // `AddBoxOverlay` always draws outlines based on the color given so `AddBoxOverlay2` is used
+        // 0.001 duration used as 0, FLT_EPSILON, 0.01, etc. flickers for some reason
+        // too high of duration causes too much to be rendered at once and can crash / lag the game
+        debugoverlay->AddBoxOverlay(vecCurr, min, Vector(iWidth, 0, m_flZoneHeight), angle, 
+            faceColor.r(), faceColor.g(), faceColor.b(),faceAlpha, mom_zone_draw_overlay_duration.GetFloat());
+    }
 }
 
 int C_BaseMomZoneTrigger::DrawModel(int flags)
 {
-    if (ShouldDrawOutline())
+    const auto pPlayer = C_MomentumPlayer::GetLocalMomPlayer();
+    if (!pPlayer)
+        return 0;
+
+    const auto pRunEntData = pPlayer->GetCurrentUIEntData();
+    if (!pRunEntData)
+        return 0;
+
+    const auto pStartZone = dynamic_cast<C_TriggerTimerStart*>(this);
+    if (!pStartZone && m_iTrackNumber != pRunEntData->m_iCurrentTrack)
```

Comment:
Not a really big fan of doing a dynamic cast every render call. A better solution here would be to network [the zone type](https://github.com/momentum-mod/game/blob/develop/mp/src/game/server/momentum/mom_triggers.cpp#L166) across to here and check that instead. You can set the zone type in the constructor of each zone trigger.

Reply:
Or, instead of networking this, you could just do the exact same code (virtual base class that all the children override and return the right value). Might be better that way, yeah.

Reply:
Did the latter